### ISSUE TITLE
IDE: possible memory leak with long running scripts

### ISSUE THREAD

**Issue by RaiMan** [oculix-org/Oculix#308]
[originally filed on Launchpad](https://bugs.launchpad.net/sikuli/+bug/1532541)

[another discussion on launchpad](https://answers.launchpad.net/sikuli/+question/691654)

**Possible workarounds (Java 11+)**

```
- start the IDE normally and run your script in the IDE

- in $JAVA_HOME/bin there is a command jcmd

- in parallel to the running script run it in a terminal as $JAVA_HOME/bin/jcmd without any parameters
   and you will get the active JVMs

- find the pid for the JVM running org.sikuli.ide.SikulixIDE

- from time to time in the same terminal window run
   jcmd <IDE-PID> GC.run

This should run a garbage collection and the memory usage should be at a lower level afterwards.
```

```
... or you might trigger the garbage collection directly inside the script:

import java.lang.System as JSYS # at beginning of main script

JSYS.gc() # at loop start or loop end

... or somewhere else, where it makes sense.
```

**Comment by RaiMan** [comment]
this is a script related to
https://answers.launchpad.net/sikuli/+question/692550

```
Reg=Region(1400+0,675,365,51)
Reg.highlight(2)

import java.lang.Runtime as RT
jrt = RT.getRuntime()
print "max:", jrt.maxMemory(), "tot:", jrt.totalMemory(), "free:", jrt.freeMemory()

for i in range(1,2000):
    Reg.text()
    if i % 50 == 0:
        print i, "free:", int(jrt.freeMemory()/(1024 * 1024)) 
```

**Comment by eight04** [comment]
Can still reproduce. Sikuli consumes 3GB RAM after 30 minutes.

My script also uses OCR.

SikulixIDE 2.0.5

**Comment by eight04** [comment]
`JSYS.gc()` doesn't work. The IDE itself still consumes 3G memory even when the script is not running.

The script executed this time doesn't use any OCR feature.

**Comment by pcarbonn** [comment]
I do observe a memory leak in the SikuliIDE, and adding `JSYS.gc()`  at various places does help.

**Comment by julienmerconsulting** [comment]
[![Repo](https://img.shields.io/badge/now_at-oculix--org%2FOculix-00B4D8?style=flat-square)](https://github.com/oculix-org/Oculix)
[![Status](https://img.shields.io/badge/legacy_bug-not_yet_fixed-f59e0b?style=flat-square)](.)
[![Latest](https://img.shields.io/badge/latest-v3.0.3-8b5cf6?style=flat-square)](https://github.com/oculix-org/Oculix/releases/tag/v3.0.3)

Hi @eight04 and @pcarbonn — taking over this thread as the new maintainers. SikuliX has transitioned to **OculiX** at https://github.com/oculix-org/Oculix, [with @RaiMan's blessing on his canonical GitBook](https://raimans-sikulix.gitbook.io/sikulix.com).

## 🔍 Honest status of this issue in 2026

We haven't specifically fixed this in OculiX yet. The leak you describe on SikuliX 2.0.5 is a legacy issue we inherited and haven't done targeted investigation on. **I can't claim "switch to OculiX and the leak is gone."**

What I *can* say is: **the memory profile is different enough in OculiX 3.0.3 that retesting is worthwhile**.

## 🆕 Why retesting on OculiX 3.0.3 might surprise you

| Change | Why it might shift the symptoms |
|---|---|
| **Java 21+ baseline** (vs SikuliX's Java 8/11) | Modern JVM uses G1GC by default with much better long-running behavior. ZGC available via `-XX:+UseZGC`. |
| **Tesseract via [Legerix](https://github.com/oculix-org/Legerix)** (bundled native) | OCR memory now controlled by Legerix's native bundle, different leak vectors than system-wide Tesseract. |
| **OpenCV via [Apertix](https://github.com/oculix-org/Apertix)** (bundled native) | Vision pipeline native memory contained in Apertix. |
| **`ScreenUnion` removed** ([`068eac18`](https://github.com/oculix-org/Oculix/commit/068eac18), May 2026, by @RaiMan) | Eliminated a memory-heavy unused legacy abstraction. |

## 🛠️ Workarounds that still apply

@RaiMan's 2020 suggestion still works in OculiX:

```python
import java.lang.System as JSYS
JSYS.gc()  # at loop start or end
```

Modern JVM flags worth trying with `oculixide-3.0.3`:

```bash
java -Xms512m -Xmx2g \
     -XX:+UseG1GC \
     -XX:+UnlockExperimentalVMOptions \
     -XX:G1PeriodicGCInterval=30000 \
     -jar oculixide-3.0.3-linux.jar
```

The `G1PeriodicGCInterval=30000` forces G1GC to do a periodic full GC every 30 seconds even when the heap isn't full — directly mitigates "memory creeping up over 30+ minutes" patterns.

## 📋 What would help us actually fix this

1. **Retest with OculiX 3.0.3** — see if your leak rate is different (the JVM modernization alone often changes the pattern dramatically)
2. **If still leaking on OculiX**, please [open a new issue on oculix-org/Oculix](https://github.com/oculix-org/Oculix/issues/new) with:
   - JVM version (`java -version`)
   - OS + RAM
   - Minimal script that reproduces
   - Heap dump from `jcmd <pid> GC.heap_dump` after the leak materializes
3. We'll prioritize it for **v3.0.4** if reproducible

Insights show ~96 unique CI environments pulling OculiX per day — long-running script stability matters for that audience. Your reproductions would directly drive the fix.

Thanks for keeping this thread alive. The transition from SikuliX to OculiX doesn't make 6 years of legitimate bug reports disappear — they stay on our backlog. 🦎
